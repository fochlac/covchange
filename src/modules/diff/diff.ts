export function diffReports(original: Core.Report, changed: Core.Report): Core.DiffReport {
	const deletedFiles = Object.keys(original.files).filter(name => !changed.files[name])
	const changedFiles = Object.keys(original.files).filter(name => {
		const altered = changed.files[name] && changed.files[name].metrics
		const orig = original.files[name].metrics
		if (!altered) {
			return false
		}

		return altered && (altered.statements !== orig.statements || altered.coveredStatements !== orig.coveredStatements)
	})
	const newFiles = Object.keys(changed.files).filter(name => !original.files[name])

	return {
		total: diffMetrics(original.metrics, changed.metrics),
		deleted: deletedFiles,
		changed: changedFiles.reduce((files, name) => {
			files[name] = diffMetrics(original.files[name].metrics, changed.files[name].metrics)
			return files
		}, {}),
		new: newFiles.reduce((files, name) => {
			files[name] = changed.files[name].metrics
			return files
		}, {}),
	}
}

function diffMetrics(original: Core.Metrics, changed: Core.Metrics): Core.Diff {
	return {
		original,
		changed,
		diff: {
			statements: Math.floor((changed.statements - original.statements) * 10) / 10,
			conditionals: Math.floor((changed.conditionals - original.conditionals) * 10) / 10,
			coveredStatements: Math.floor((changed.coveredStatements - original.coveredStatements) * 10) / 10,
			coveredConditionals: Math.floor((changed.coveredConditionals - original.coveredConditionals) * 10) / 10,
			statementCov: Math.floor((changed.statementCov - original.statementCov) * 10) / 10,
			conditionalCov: Math.floor((changed.conditionalCov - original.conditionalCov) * 10) / 10,
		},
	}
}
