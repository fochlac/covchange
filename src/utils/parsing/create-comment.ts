const svgBaseUrl = `http://${global.address}:${global.port}/svg`

const symbolsMap = {
	critical: `![critical](${svgBaseUrl}/critical.svg)`,
	low: `![warning](${svgBaseUrl}/warning.svg)`,
	medium: `![ok](${svgBaseUrl}/ok.svg)`,
	high: `![good](${svgBaseUrl}/good.svg)`,
	optimal: `![top](${svgBaseUrl}/top.svg)`,
	decrease: `![decrease](${svgBaseUrl}/arrow-down.svg)`,
	increase: `![improvement](${svgBaseUrl}/arrow-up.svg)`,
	new: `![new](${svgBaseUrl}/new.svg)`,
}

export function createCommentObject(diff: Core.DiffReport, comment?: Core.Comment): Core.CommentRest {
	const totalDiff = diff.total.diff.statementCov
	const changedFilesHeader = (Object.keys(diff.changed).length && ['| Quality | File | Change | Coverage |', '|---|---|---|---|']) || []
	const deletedFiles = diff.deleted.length ? ['##### Deleted files', '```diff', ...diff.deleted.map(name => `- ${name}`), '```'] : []
	const covSymbol = totalDiff < 0 ? symbolsMap.decrease : symbolsMap.increase
	const summedMetrics = Object.values(diff.changed)
		.map(diff => diff.changed)
		.concat(Object.values(diff.new))
		.reduce(
			(sum, metric) => ({
				coveredStatements: sum.coveredStatements + metric.coveredStatements,
				statements: sum.statements + metric.statements,
			}),
			{ coveredStatements: 0, statements: 0 },
		)
	const diffCoverage = Math.round((summedMetrics.coveredStatements / summedMetrics.statements) * 100)

	const lines = [
		'### Coverage Statistics',
		`#### ${covSymbol} This pull request has a diff coverage of ${diffCoverage}% and will ${totalDiff < 0 ? 'decrease' : 'increase'}` +
			` total coverage by ${Math.abs(totalDiff)}% to ${diff.total.changed.statementCov}%.`,
		'',
		...changedFilesHeader,
		...Object.keys(diff.changed)
			.sort((a, b) => diff.changed[a].changed.statementCov - diff.changed[b].changed.statementCov)
			.slice(0, 10)
			.map(name => changedFileInfo(name, diff.changed[name])),
		...Object.keys(diff.new)
			.sort((a, b) => diff.new[a].statementCov - diff.new[b].statementCov)
			.slice(0, 10)
			.map(name => newFileInfo(name, diff.new[name])),
		...deletedFiles,
	]

	return { text: lines.join('\n'), version: comment && comment.version }
}

function newFileInfo(name: string, metrics: Core.Metrics) {
	const newCoverageCell = `${Math.round(metrics.statementCov)}% (${metrics.coveredStatements}/${metrics.statements})`

	return `| ${getSymbolFromCoverage(metrics)} | ${name} | ${symbolsMap.new} *new* | ${newCoverageCell} |`
}

function changedFileInfo(name: string, { diff: { statementCov }, changed, changed: { coveredStatements }, original }: Core.Diff) {
	const covSymbol = statementCov < 0 ? symbolsMap.decrease : symbolsMap.increase
	const diffPercentage = `${statementCov < 0 ? '-' : '+'}${Math.abs(Math.round(statementCov))}%`
	const diffCell = `${covSymbol} ${diffPercentage} (${original.coveredStatements} :arrow_right: ${coveredStatements})`
	const newCoverageCell = `${Math.round(changed.statementCov)}% (${coveredStatements}/${changed.statements})`

	return `| ${getSymbolFromCoverage(changed)} | ${name} | ${diffCell} | ${newCoverageCell} |`
}

const getSymbolFromCoverage = (metrics: Core.Metrics) => {
	const { statementCov, statements, conditionalCov, conditionals } = metrics
	let stringcode = 'critical'
	if (statementCov >= 90 && (conditionalCov > 75 || conditionals === 0)) {
		stringcode = 'optimal'
	} else if (statementCov >= 80 && (conditionalCov > 50 || conditionals <= 1)) {
		stringcode = 'high'
	} else if (statementCov >= 70 || statements <= 5) {
		stringcode = 'medium'
	} else if (statementCov > 50) {
		stringcode = 'low'
	}

	return symbolsMap[stringcode]
}
