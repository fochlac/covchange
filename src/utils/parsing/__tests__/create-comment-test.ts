import { mockReport, mockReport2 } from '../../../__tests__/data/reports'

import { createCommentObject } from '../create-comment'
import { diffReports } from '../diff'

jest.mock('../../../utils/logger.ts', () => () => null)

describe('diff-module', () => {
	const diff = diffReports(mockReport, mockReport2)

	it('should create a comment from the diff report', () => {
		const comment = createCommentObject(diff)
		expect(comment.version).toBeUndefined()
		expect(comment).toMatchSnapshot()
	})

	it('should create a comment from the diff report', () => {
		const comment = createCommentObject(diff, { id: '0', commentId: '0', version: 1 })
		expect(comment.version).toEqual(1)
	})

	it('should create a comment from the diff report', () => {
		const comment = createCommentObject(diff, { id: '0', commentId: '0', version: 1 }, 'http://ttbsbld102.dev.ttw:8080/job/Atlantis-PR/298936/PR-UnitTestCoverage')
		expect(comment).toMatchSnapshot()
	})
})
