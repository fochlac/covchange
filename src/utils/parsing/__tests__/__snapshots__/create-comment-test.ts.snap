// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff-module should create a comment from diff of equal reports 1`] = `
Object {
  "text": "### Coverage Statistics
#### No Changes in Coverage",
  "version": undefined,
}
`;

exports[`diff-module should create a comment from the diff report 1`] = `
Object {
  "text": "### Coverage Statistics
#### ![improvement](http://undefined:undefined/svg/arrow-up.svg) This pull request has a diff coverage of 84% and will increase total coverage by 2.38% to 87.19%.

| Quality | File | Change | Statements | Conditionals |
|---|---|---|---|---|
| ![critical](http://undefined:undefined/svg/critical.svg) | orders/app/index.js | ![improvement](http://undefined:undefined/svg/arrow-up.svg) +8% (0 :arrow_right: 1) | 8% (1/13) | 0% (0/4) |
| ![good](http://undefined:undefined/svg/good.svg) | orders/app/components/order-tours-tab/order-tours-table.js | ![improvement](http://undefined:undefined/svg/arrow-up.svg) +81% (0 :arrow_right: 22) | 81% (22/27) | 85% (35/41) |
| ![top](http://undefined:undefined/svg/top.svg) | orders/app/components/orders-view.js | ![decrease](http://undefined:undefined/svg/arrow-down.svg) -6% (47 :arrow_right: 44) | 94% (44/47) | 100% (17/17) |
| ![ok](http://undefined:undefined/svg/ok.svg) | orders/app/components/orders-filter-bar.js | ![improvement](http://undefined:undefined/svg/arrow-up.svg) +13% (56 :arrow_right: 53) | 96% (53/55) | 50% (11/22) |
| ![top](http://undefined:undefined/svg/top.svg) | orders/app/components/order-tours-tab/order-tours-table-renderer.js | ![improvement](http://undefined:undefined/svg/arrow-up.svg) +98% (0 :arrow_right: 44) | 98% (44/45) | 94% (33/35) |
| ![warning](http://undefined:undefined/svg/warning.svg) | orders/app/components/order-tours-tab/order-tours-table-core.js | ![new](http://undefined:undefined/svg/new.svg) *new* | 55% (6/11) | 71% (5/7) |
| ![warning](http://undefined:undefined/svg/warning.svg) | orders/app/components/orders-subviews.js | ![new](http://undefined:undefined/svg/new.svg) *new* | 68% (17/25) | 95% (18/19) |
##### Deleted files
\`\`\`diff
- orders/module.js
\`\`\`",
  "version": undefined,
}
`;

exports[`diff-module should create a comment from the diff report 2`] = `
Object {
  "text": "### Coverage Statistics
#### ![improvement](http://undefined:undefined/svg/arrow-up.svg) This pull request has a diff coverage of 84% and will increase total coverage by 2.38% to 87.19%.

| Quality | File | Change | Statements | Conditionals |
|---|---|---|---|---|
| ![critical](http://undefined:undefined/svg/critical.svg) | [orders/app/index.js](http://ttbsbld102.dev.ttw:8080/job/Atlantis-PR/298936/PR-UnitTestCoverage/orders/app/index.js.html) | ![improvement](http://undefined:undefined/svg/arrow-up.svg) +8% (0 :arrow_right: 1) | 8% (1/13) | 0% (0/4) |
| ![good](http://undefined:undefined/svg/good.svg) | [orders/app/components/order-tours-tab/order-tours-table.js](http://ttbsbld102.dev.ttw:8080/job/Atlantis-PR/298936/PR-UnitTestCoverage/orders/app/components/order-tours-tab/order-tours-table.js.html) | ![improvement](http://undefined:undefined/svg/arrow-up.svg) +81% (0 :arrow_right: 22) | 81% (22/27) | 85% (35/41) |
| ![top](http://undefined:undefined/svg/top.svg) | [orders/app/components/orders-view.js](http://ttbsbld102.dev.ttw:8080/job/Atlantis-PR/298936/PR-UnitTestCoverage/orders/app/components/orders-view.js.html) | ![decrease](http://undefined:undefined/svg/arrow-down.svg) -6% (47 :arrow_right: 44) | 94% (44/47) | 100% (17/17) |
| ![ok](http://undefined:undefined/svg/ok.svg) | [orders/app/components/orders-filter-bar.js](http://ttbsbld102.dev.ttw:8080/job/Atlantis-PR/298936/PR-UnitTestCoverage/orders/app/components/orders-filter-bar.js.html) | ![improvement](http://undefined:undefined/svg/arrow-up.svg) +13% (56 :arrow_right: 53) | 96% (53/55) | 50% (11/22) |
| ![top](http://undefined:undefined/svg/top.svg) | [orders/app/components/order-tours-tab/order-tours-table-renderer.js](http://ttbsbld102.dev.ttw:8080/job/Atlantis-PR/298936/PR-UnitTestCoverage/orders/app/components/order-tours-tab/order-tours-table-renderer.js.html) | ![improvement](http://undefined:undefined/svg/arrow-up.svg) +98% (0 :arrow_right: 44) | 98% (44/45) | 94% (33/35) |
| ![warning](http://undefined:undefined/svg/warning.svg) | [orders/app/components/order-tours-tab/order-tours-table-core.js](http://ttbsbld102.dev.ttw:8080/job/Atlantis-PR/298936/PR-UnitTestCoverage/orders/app/components/order-tours-tab/order-tours-table-core.js.html) | ![new](http://undefined:undefined/svg/new.svg) *new* | 55% (6/11) | 71% (5/7) |
| ![warning](http://undefined:undefined/svg/warning.svg) | [orders/app/components/orders-subviews.js](http://ttbsbld102.dev.ttw:8080/job/Atlantis-PR/298936/PR-UnitTestCoverage/orders/app/components/orders-subviews.js.html) | ![new](http://undefined:undefined/svg/new.svg) *new* | 68% (17/25) | 95% (18/19) |
##### Deleted files
\`\`\`diff
- orders/module.js
\`\`\`",
  "version": 1,
}
`;
